core/lib/dal/src/tee_proof_generation_dal.rs

```diff
@@ -1,122 +1,110 @@
 use std::time::Duration;
 
-use strum::{Display, EnumString};
 use zksync_db_connection::{
-    connection::Connection,
-    error::DalResult,
-    instrument::{InstrumentExt, Instrumented},
+    connection::Connection, error::DalResult, instrument::Instrumented,
     utils::pg_interval_from_duration,
 };
 use zksync_types::{tee_types::TeeType, L1BatchNumber};
 
-use crate::Core;
+use crate::{
+    models::storage_tee_proof::{StorageTeeProof, TmpStorageTeeProof},
+    Core,
+};
 
 #[derive(Debug)]
 pub struct TeeProofGenerationDal<'a, 'c> {
     pub(crate) storage: &'a mut Connection<'c, Core>,
 }
 
-#[derive(Debug, EnumString, Display)]
-enum TeeProofGenerationJobStatus {
-    #[strum(serialize = "ready_to_be_proven")]
-    ReadyToBeProven,
-    #[strum(serialize = "picked_by_prover")]
-    PickedByProver,
-    #[strum(serialize = "generated")]
-    Generated,
-    #[strum(serialize = "skipped")]
-    Skipped,
-}
-
 impl TeeProofGenerationDal<'_, '_> {
-    pub async fn get_next_block_to_be_proven(
+    pub async fn get_next_batch_to_be_proven(
         &mut self,
+        tee_type: TeeType,
         processing_timeout: Duration,
-    ) -> DalResult<Option<L1BatchNumber>> {
+    ) -> DalResult<Option<(i64, L1BatchNumber)>> {
```

Comment:
Why do you need to expose IDs? Batch number and TEE type should form a unique identifier of the proof; why do we need something else?

Reply:
Good catch!

In this PR, I initially only wanted to set up an RPC method to get SGX proofs and attestations. However, @haraldh was concerned that my endpoint only returns a single proof for each batch number, so I ended up redesigning the SQL schema for the SGX proofs. This made the PR much bigger in scope than I originally expected.

Harald suggested it might be better if we could submit _multiple_ proofs per `(batch_number, tee_type)` (_not_ just for different TEE types). The main reason is that SGX attestations may not be valid forever; they can be revoked, and so on, so we might need multiple instances. This could apply to other types of TEEs as well.

If you have any suggestions for a better implementation, feel free to share! I can also revert to a single proof per `(batch_number, tee_type)` if you think it's better to introduce complexity to the system more gradually.

Also, thanks for checking things out while it's still a work in progress! It could save a lot of time if I'm heading in the wrong direction with this PR.

Reply:
Yes, probably — I don't see how it contradicts what I wrote, as I still need to be able to store multiple instances of proofs for a single `(batch_number, tee_type)` pair. We need a way to differentiate between them so that the prover knows which row in the `tee_proof` SQL table to update with the proof.

It's very nuanced when it comes to implementation, so let me briefly summarize how it works to help us be more precise when discussing implementation details.

## Old implementation

The simplified implementation of prover-node communication from the `main` branch:
1. The prover calls the `/tee/proof_inputs` endpoint, which triggers `get_next_batch_to_be_proven` to return the `L1BatchNumber` of the batch to be proven, [along](https://github.com/matter-labs/zksync-era/blob/2fa6bf0ffa5d7a5ff62d595a0efeff9dcd9e5a1a/core/node/proof_data_handler/src/tee_request_processor.rs#L60-L68) with the `TeeVerifierInput` for that batch number.
   https://github.com/matter-labs/zksync-era/blob/c9f76e571e5570d2c4194feee03bb260b24c378f/core/lib/dal/src/tee_proof_generation_dal.rs#L32-L35
   The batch `status` is set to `picked_by_prover` in the database to ensure that the next call to `/tee/proof_inputs` does not return the same batch:
   https://github.com/matter-labs/zksync-era/blob/c9f76e571e5570d2c4194feee03bb260b24c378f/core/lib/dal/src/tee_proof_generation_dal.rs#L41
1. The prover computes the proof and saves it back in the database by calling `/tee/submit_proofs/{l1_batch_number}`, which internally triggers `save_proof_artifacts_metadata`:
   https://github.com/matter-labs/zksync-era/blob/c9f76e571e5570d2c4194feee03bb260b24c378f/core/lib/dal/src/tee_proof_generation_dal.rs#L80-L106
   Using `l1_batch_number` as the unique ID for submitting a proof isn't appropriate here because we want to allow multiple proofs per `l1_batch_number`, even for the same TEE type. That's why I introduced a new implementation in this PR.

## New implementation

The simplified implementation introduced in this PR:
1. The prover calls the `/tee/proof_inputs` endpoint with a `tee_type` payload to retrieve the smallest batch number that has not yet been proved for that `tee_type`, [along](https://github.com/matter-labs/zksync-era/blob/2fa6bf0ffa5d7a5ff62d595a0efeff9dcd9e5a1a/core/node/proof_data_handler/src/tee_request_processor.rs#L60-L68) with the `TeeVerifierInput` for that batch number. Under the hood, this triggers `get_next_batch_to_be_proven`, which returns the row's `ID` (i64) and the `L1BatchNumber` to be proven.
   https://github.com/matter-labs/zksync-era/blob/cea2cf98860a65a7ffcb34094833e344874be8e9/core/lib/dal/src/tee_proof_generation_dal.rs#L20-L24
   Under the hood [we set](https://github.com/matter-labs/zksync-era/blob/c9f76e571e5570d2c4194feee03bb260b24c378f/core/lib/dal/src/tee_proof_generation_dal.rs#L43) `prover_taken_at` to indicate that this `(id, l1_batch_number)` is currently being processed by the prover.
1. The prover computes the proof and saves it in the database by calling `/tee/submit_proofs/{l1_batch_number}` with the `proof_id` in the payload, indicating which row to update with the proof. See the details here:
    https://github.com/matter-labs/zksync-era/blob/cea2cf98860a65a7ffcb34094833e344874be8e9/core/lib/dal/src/tee_proof_generation_dal.rs#L71-L93
   This is the main reason, in response to your original comment, why I return `DalResult<Option<(i64, L1BatchNumber)>>` from `get_next_batch_to_be_proven` instead of `DalResult<Option<L1BatchNumber>>`.

Reply:
For the record: after an offline discussion with @popzxc, we decided to have only a single proof per TEE type per batch. A more extensive redesign will be needed before transitioning to a more flexible system that allows multiple proofs per TEE type per batch. cc @haraldh @thomasknauth